### ISSUE TITLE
Jupyter Notebook Support for Legion Python

### ISSUE THREAD

**Issue by lightsighter** [StanfordLegion/legion#801]
The last remaining TODO for #726 is to add support for Jupyter notebooks for Legion Python.

**Comment by lightsighter** [comment]
Work on new equivalence sets has delayed this to the next milestone.

**Comment by eddy16112** [comment]
Jupyter notebook support for single node execution or multi-node execution without control replication has been added to the master branch.

**Comment by lightsighter** [comment]
Still leaving this issue open so we can discuss multi-node execution with control replication, but definitely very good progress.

**Comment by elliottslaughter** [comment]
Is there anything to still do here?

**Comment by lightsighter** [comment]
Yes, there is still the "multi-node" execution with control replication feature which we should probably do for scalability reasons (and because Legate might need it).

**Comment by lightsighter** [comment]
Documenting the control replication feature here for @manopapad. The idea is to implement a custom Jupyter notebook kernel that rather than directly forwarding commands and data directly to the Python interpreter will instead distribute this using a tree broadcast out to however many top-level tasks are running in a control replicated fashion across a large machine. Each top-level task will effectively be listening for these commands and then interpret them in the Python interpreter on its local node in a control-replicated sense. This will be much more scalable than running in a non-control-replicated fashion and having tasks dispatched from a single controller node. This will require creating a custom Jupyter notebook kernel. It would also be good to use Realm directly to do the broadcast rather than relying on slow messaging layers like ZeroMQ.
